### PR TITLE
Mav 3943/intercom

### PR DESCRIPTION
**Pull request by AlexBelz123**
# Security Diff Notes

Purpose: quick debugging context for the security changes implemented from `security-tickets/security-investigation-report.md`.

This file explains what changed, which ticket it belongs to, and why the change exists. It is not a full investigation report.

## Implementation Order Used

1. SEC-06: intercom port classification and privileged request gating.
2. SEC-10: top-frame-only dApp bridge.
3. SEC-03: inherited ad-injection removal.
4. SEC-12: local-first token clearing and token response validation.
5. SEC-08: strict auth challenge contract before signing.
6. SEC-07: remove raw password hash session storage.
7. SEC-09/SEC-11: documented as backend-blocked because the Go backend repo was not present.

### What Changed

- `src/lib/intercom/server.ts`
  - Added `IntercomPortKind`: `extension-ui`, `confirm-ui`, `content-script-relay`, `unknown`.
  - Replaced the plain `Set<Runtime.Port>` with `Map<Runtime.Port, IntercomPortInfo>`.
  - Added `classifyIntercomPort()`, `getPortInfo()`, and helper predicates.
  - Request handlers now receive `(payload, port, portInfo)`.

- `src/lib/intercom/index.ts`
  - Re-exported the new intercom role types and helpers.

- `src/lib/temple/back/intercom-permissions.ts`
  - New Temple-specific request policy helper.
  - Content-script relay ports may only send:
    - `TempleMessageType.PageRequest`
    - `TempleMessageType.Acknowledge`
  - Relay ports with known nonzero `frameId` are rejected even for relay-safe requests.
  - Unknown ports are rejected.
  - Extension UI and confirm UI are trusted for normal background UI requests.

- `src/lib/temple/back/main.ts`
  - Applies `assertTempleRequestAllowedForPortInfo()` before dispatching any `TempleRequest`.

- `src/lib/temple/back/actions.ts`
  - Internal wallet-originated `OperationsRequest` and `SignRequest` now require an `ExtensionUi` port.
  - Their `ConfirmationRequest` responses must also come from that same trusted extension UI port.

- `src/lib/temple/back/dapp.ts`
  - DApp confirmation payload fetch/approval now requires `ConfirmUi`.
  - This prevents relay ports or normal page/content-script ports from fetching or resolving dApp confirmation payloads.

### Why

Before this change, content scripts and extension UI used the same `INTERCOM` port name and both carried the extension runtime id. The background accepted privileged requests from both. A content-script context could therefore attempt privileged internal requests while the wallet was unlocked.

The fix makes port provenance explicit and gates privileged message types by role.

### Debugging Notes

- If popup/fullpage requests fail with `Unauthorized intercom request source`, inspect `port.sender.url`, `port.sender.tab`, and the classification in `classifyIntercomPort()`.
- If dApp confirmation windows fail, check that the sender URL resolves to `confirm.html`.
- If a legitimate top-level dApp stops working, inspect relay `frameId`; `frameId: 0` or missing is allowed, nonzero is rejected.

### Tests

- `src/lib/intercom/server.test.ts`
- `src/lib/temple/back/intercom-permissions.test.ts`

### What Changed

- `webpack/manifest.ts`
  - `scripts/contentScript.js` now has `all_frames: false`.
  - `scripts/keepBackgroundWorkerAlive.js` remains `all_frames: true`; it is not the dApp bridge.

- `src/contentScript.ts`
  - Added `isTopFrame()` using `window.top === window`.
  - External-link analytics runs only in the top frame.
  - The `window.message` dApp/Beacon bridge listener is registered only in the top frame.

### Why

Previously the bridge listener was available in every injected iframe. A malicious iframe embedded in a trusted page could access the wallet bridge under the iframe origin, creating origin-confusion and phishing risk.

### Debugging Notes

- Iframe-embedded dApps are intentionally blocked.
- If a top-level dApp no longer receives responses, first verify the page is not running inside an iframe.
- The generated manifest should show `contentScript.js` with `all_frames: false`.

### What Changed

- `webpack.config.ts`
  - Removed the `replaceAds` script entry.

- `webpack/manifest.ts`
  - Removed the `scripts/replaceAds.js` content script.
  - Removed `misc/ad-banners/*` from web-accessible resources.

- `src/background.ts`
  - Removed startup `updateRulesStorage()` call.

- `src/lib/temple/back/main.ts`
  - Removed `ContentScriptType.UpdateAdsRules` handling.
  - Removed `ContentScriptType.ExternalAdsActivity` handling.

- `src/lib/constants.ts`
  - Removed ad-injection-only constants:
    - `ContentScriptType.ExternalAdsActivity`
    - `ContentScriptType.UpdateAdsRules`
    - `ALL_ADS_RULES_STORAGE_KEY`
    - `ADS_RULES_UPDATE_INTERVAL`
    - `TEMPLE_WALLET_AD_ATTRIBUTE_NAME`
  - Preserved in-app partner promotion constants such as `AD_HIDING_TIMEOUT` and `AD_SEEN_THRESHOLD`.

- `src/lib/apis/temple/index.ts`
  - Removed the ad-rules endpoint barrel export.

- `src/lib/env.ts`
  - Removed unused `TEMPLE_WALLET_DEVELOPMENT_BRANCH_NAME`.
  - Preserved `TEMPLE_WALLET_API_URL` because it is used by non-ad Temple API consumers.

- Deleted:
  - `src/replaceAds.ts`
  - `src/content-scripts/replace-ads/**`
  - `src/lib/ads/update-rules-storage.ts`
  - `src/lib/apis/temple/endpoints/ad-rules.ts`
  - `public/misc/ad-banners/tkey-inpage-ad.png`

### Why

The inherited Temple ad subsystem modified third-party pages and fetched Temple/Madfish ad rules from the background. This was unnecessary for Mavryk Wallet and increased privacy, store-policy, performance, and page-mutation risk.

### Debugging Notes

- If build fails with missing `lib/ads/update-rules-storage` or `replaceAds`, there is a stale import.
- If in-app partner promotion breaks, do not restore the deleted page-injection path. Check in-app promotion imports first.
- Expected built scripts:
  - `scripts/contentScript.js`
  - `scripts/keepBackgroundWorkerAlive.js`
  - no `scripts/replaceAds.js`

### What Changed

- `src/mavryk/api/storage.ts`
  - Added helpers to collect stored refresh tokens and clear auth tokens independently:
    - `collectStoredRefreshTokens()`
    - `clearStoredAuthTokens()`
  - `clearAllAuthTokensFromStorage()` now clears scoped tokens for the wallet across networks, plus legacy global keys.

- `src/mavryk/api/auth.ts`
  - `logoutAuth()` now:
    1. Captures stored refresh tokens.
    2. Clears local auth tokens first.
    3. Calls `/auth/logout` best-effort with captured refresh tokens.
    4. Logs remote logout failures but does not restore local tokens.

- `src/mavryk/api/auth.schema.ts`
  - Access tokens must be JWT-like (`header.payload.signature`).
  - Access tokens must have a parseable, unexpired `exp`.
  - Refresh tokens must be non-empty and at most 4096 characters.

- `src/mavryk/api/client.ts`
  - Auth endpoints do not get an `Authorization` header attached:
    - `/auth/challenge`
    - `/auth/verify`
    - `/auth/refresh`
    - `/auth/logout`

### Why

Lock previously cleared local tokens only after successful remote logout. If the network/backend failed, plaintext local tokens could remain. The new behavior makes lock/logout local-first.

Token response validation prevents malformed or expired access tokens from entering storage.

### Debugging Notes

- After `lock()`, check storage before debugging backend logout: local tokens should already be gone.
- Logout errors are expected to be logged if the backend is unreachable.
- If auth starts failing immediately after verify/refresh, inspect token `exp` and JWT shape.
- If a backend returns an access token without `exp`, the frontend now rejects it.

### Tests

- `src/mavryk/api/auth.test.ts`
- `src/mavryk/api/client.test.ts`

### What Changed

- `src/mavryk/api/auth-payload.helpers.ts`
  - Added structured auth challenge construction:
    - `buildStructuredAuthChallengeMessage()`
    - `buildStructuredAuthChallengePayloadHex()`
  - Added `validateAuthChallengeForSigning()`.
  - Required fields:
    - fixed `Mavryk Wallet Authentication` prefix
    - wallet address
    - network id
    - nonce
    - ISO expiry
    - audience `mavryk-wallet-api`
  - Enforced:
    - max challenge length: 1024
    - nonce pattern: `^[A-Za-z0-9._~-]{16,256}$`
    - expiry in the future
    - expiry no more than 10 minutes ahead
    - exact match between backend `challenge` and locally reconstructed message

- `src/mavryk/api/auth.schema.ts`
  - Challenge response schema now validates nonce shape and challenge length.

- `src/mavryk/api/auth.ts`
  - `requestAuthChallenge()` validates the challenge before storing it as the last challenge/nonce.

- `src/mavryk/api/utils.ts`
  - `signAuthChallengeWithVault()` no longer signs arbitrary legacy text.
  - It only signs structured auth challenge messages.

- `docs/api/auth.md`
  - Documents the exact structured challenge format and client validation rules.

### Why

Before this change, the backend controlled arbitrary challenge text and the wallet silently signed it during unlock/import auth. That created risk if the backend was compromised or returned a replayable/domain-confused payload.

The client now only signs a strict, domain-separated auth message that binds wallet address, network, nonce, expiry, and audience.

### Debugging Notes

- Backend must return exactly this message shape:

```text
Mavryk Wallet Authentication

Please sign this message to authenticate.

Wallet Address: <mv1...>
Network: <mainnet|basenet|...>
Nonce: <nonce>
Expires At: <ISO timestamp>
Audience: mavryk-wallet-api
```

- Any text mismatch fails before storage/signing.
- Common failures:
  - nonce shorter than 16 chars
  - expiry too far in the future
  - expiry string lacks ISO timestamp shape
  - backend uses selected account address while frontend signs with auth wallet address
  - network id mismatch after network normalization

### Tests

- `src/mavryk/api/auth-payload.helpers.test.ts`
- `src/mavryk/api/utils.test.ts`
- `src/mavryk/api/auth.test.ts`

### What Changed

- `src/lib/temple/back/vault/session-store.ts`
  - Replaced raw `@Vault:session.passHash` storage with versioned payload key `@Vault:session.v2`.
  - New payload includes:
    - `version`
    - `createdAt`
    - `expiresAt`
    - random `wrappingKey`
    - AES-GCM `iv`
    - `wrappedPassHash`
  - Removes legacy `@Vault:session.passHash` when encountered.
  - Rejects and clears invalid/tampered/expired session payloads.
  - Calls `browser.storage.session.setAccessLevel({ accessLevel: 'TRUSTED_CONTEXTS' })` when available.

- `src/lib/temple/back/vault/index.ts`
  - `Vault.setup()` saves wrapped session material when MV3 worker session recovery is enabled.
  - `Vault.recoverFromSession()` unwraps session material, imports the pass key, and validates it by decrypting `checkStrgKey`.
  - Invalid session material is cleared and ignored.
  - `Vault.forgetSession()` clears both legacy and v2 session keys.
  - `Vault.spawn()` clears session material through the new helper.

### Why

The previous MV3 recovery stored raw `SHA-256(password)` in `browser.storage.session`. That value was password-equivalent for vault decryption and offline-crackable.

The new design keeps MV3 service-worker recovery but removes the raw password hash from storage and adds versioning, expiry, trusted-context access where supported, and validation before recovery.

### Debugging Notes

- This still stores session-equivalent material during the browser session. It prevents raw password-hash recovery, not every possible extension-context session hijack.
- If users are unexpectedly locked after service-worker restart:
  - inspect `@Vault:session.v2`
  - check expiry
  - check AES-GCM unwrap errors
  - check whether `browser.storage.session` is available in that browser
- Legacy `@Vault:session.passHash` should be deleted automatically and not used for unlock.

### Tests

- `src/lib/temple/back/vault/session-store.test.ts`

### What Changed

- No backend production code was changed because the Go backend repository was not present in this workspace.
- `security-tickets/` was added to `.gitignore`, so local ticket/report files stay local-only.

### Why

SEC-09 and SEC-11 require backend route authorization and lazy-add write hardening. These cannot be implemented honestly in this frontend extension repo.

### Required Backend Work

- SEC-11:
  - Authorize requested wallet addresses against the JWT principal before wallet route handlers read data.
  - Apply to history, token, RWA, portfolio, and any wallet-scoped routes.

- SEC-09:
  - Run lazy-add only after requested-wallet authorization succeeds.
  - Use checksum-valid address parsing.
  - Add unique-address quotas/rate limits.
  - Ensure unauthorized or invalid addresses never create durable wallet-index rows.

### Frontend Follow-Up After Backend Contract

- If backend uses strict `pathAddress === jwt.wallet_address`, frontend must stop using HD index-0 auth tokens for child-account protected reads.
- If backend uses authorized address sets, align `buildAuthWalletAddressesMap()` and token storage with backend registration/proof semantics.
- If watch-only reads become public, use explicit public endpoints.
- Add frontend tests so backend `403` and `429` responses are surfaced clearly, not silently treated as empty data.

## Manual QA Notes

Manual extension testing should cover:

- unlock/lock
- backend auth with structured challenge
- token clearing on lock even when logout fails
- service-worker restart session recovery
- top-level dApp connect/sign/operation
- iframe dApp rejection
- internal wallet confirmation overlay
- dApp `confirm.html` payload/approval flow
- absence of `replaceAds.js` and ad-rule requests

## Validation Commands Used

Targeted tests:

```sh
yarn test src/lib/intercom/server.test.ts src/lib/temple/back/intercom-permissions.test.ts src/lib/temple/back/vault/session-store.test.ts src/mavryk/api/auth-payload.helpers.test.ts src/mavryk/api/utils.test.ts src/mavryk/api/auth.test.ts src/mavryk/api/client.test.ts --runInBand
```

Build:

```sh
yarn build
```

Manifest/source checks:

```sh
grep -n "contentScript\\|replaceAds\\|ad-banners\\|all_frames" dist/chrome_unpacked/manifest.json
find dist/chrome_unpacked/scripts -maxdepth 1 -type f -print | sort
grep -RIn "replaceAds\\|UpdateAdsRules\\|ExternalAdsActivity\\|ALL_ADS_RULES_STORAGE_KEY\\|ADS_RULES_UPDATE_INTERVAL\\|TEMPLE_WALLET_AD_ATTRIBUTE_NAME\\|ad-rules\\|ad-banners\\|TEMPLE_WALLET_DEVELOPMENT_BRANCH_NAME" src webpack public .env.dist .env.local 2>/dev/null
```

Expected:

- Targeted Jest tests pass.
- `yarn build` passes.
- Manifest has `contentScript.js` with `all_frames: false`.
- `replaceAds.js` is absent.
- No source references remain for deleted ad-injection constants/files.

## Known Validation Caveats

- `yarn ts` may fail before project code because pinned TypeScript `4.5.5` cannot parse newer dependency `.d.ts` syntax in `node_modules`.
- `yarn lint` may report unrelated existing failures in files outside this security change set.

### CODE REVIEW

File: src/lib/temple/back/actions.ts
Comment (0xVoronov) [inline src/lib/temple/back/actions.ts]: Lets maybe remove these debug logs?

File: webpack/manifest.ts [hunk 1]
```diff
@@ -18,8 +18,6 @@ const WEB_ACCCESSIBLE_RESOURSES = [
   'scripts/*.embed.js',
   // For triggering extension page open from scripts
   'fullpage.html',
```
Comment (0xVoronov) [inline webpack/manifest.ts]: What do you think about removing fullpage.html from the WAR to prevent a case where the malicious website could embed it as an iframe on their site?

File: src/lib/temple/back/main.ts [hunk 2]
```diff
@@ -317,18 +324,9 @@ const getCurrentAccountPkh = async (): Promise<string | undefined> => {
...
-};
-
 browser.runtime.onMessage.addListener(async msg => {
   try {
     switch (msg?.type) {
-      case ContentScriptType.UpdateAdsRules:
-        await updateRulesStorage();
-        return;
       case E2eMessageType.ResetRequest:
         return clearAsyncStorages().then(() => ({ type: E2eMessageType.ResetResponse }));
```
Comment (0xVoronov) [inline src/lib/temple/back/main.ts:334]: This handler is also built into the production environment. I think we can disable it based on an environment flag and allow it only when NODE_ENV !== 'production'.

File: src/lib/temple/back/vault/session-store.ts [hunk 3]
```diff
@@ -1,41 +1,158 @@
...
+    const iv = crypto.getRandomValues(new Uint8Array(12));
+    const key = await importWrappingKey(wrappingKey);
+    const wrappedPassHash = await crypto.subtle.encrypt({ name: 'AES-GCM', iv }, key, passHashBuffer);
+    const payload: VaultSessionPayloadV2 = {
+      version: SESSION_PAYLOAD_VERSION,
+      createdAt,
+      expiresAt: createdAt + SESSION_PAYLOAD_TTL_MS,
+      wrappingKey: bytesToHex(wrappingKey),
+      iv: bytesToHex(iv),
+      wrappedPassHash: bytesToHex(wrappedPassHash)
```
Comment (0xVoronov) [inline src/lib/temple/back/vault/session-store.ts:55]: SEC-07:

I think we need to fix this. Currently, we store both the key and the PassHash in the session. If a malicious actor gains access to the session, they will get both the unsealed password and the key.

Since the password is stored unsealed and we only use `SHA-256(password)`, the attacker may be able to guess the password.

As an option, let’s store `wrappingKey` and `wrappedPassHash` separately.

I suggest storing the random `wrappingKey` in IndexedDB so that it survives a browser restart, while keeping the `wrappedPassHash` blob in the session.

Could you check the possibility of implementing this solution using `unwrapKey` (decrypt + importKey)?

**IndexedDB (disk, survives browser restarts):**

* `W` - a random AES-GCM key, non-extractable
* Usages: `['encrypt', 'unwrapKey']`, without `decrypt`

**`browser.storage.session` (memory, cleared when the browser closes):**

```text
{ version: 3, expiresAt, iv, blob = AES-GCM_W(passHash) }
```

When unlocking with the password (`Vault.setup`), we already have the raw `passHash` available (it is calculated by `toValidPassKey`). We encrypt it with key `W` using `crypto.subtle.encrypt` and store the resulting blob in `storage.session`. The raw `passHash` is not persisted anywhere else.

After a service worker restart (`recoverFromSession`), we retrieve `W` from IndexedDB and the blob from the session, then call `unwrapKey`. This is an atomic operation that decrypts the data and immediately imports it as a key inside the crypto engine:

```js
const passKey = await crypto.subtle.unwrapKey(
  'raw', blob, W,
  { name: 'AES-GCM', iv },
  { name: 'PBKDF2' },
  false,                          // non-extractable
  ['deriveBits', 'deriveKey']
);
```

The decrypted bytes never appear in JavaScript at any point. `unwrapKey` directly returns an opaque handle to the PBKDF2 key, identical to the one currently created by `Passworder.importKey(passHash)`.

After that, we can simply do `new Vault(passKey)` - the rest of the vault code remains unchanged.

The main advantage of `unwrapKey` over a simple `decrypt` + `importKey` approach is that the decrypted password/hash never gets exposed to JavaScript.

File: src/lib/temple/back/vault/session-store.ts [hunk 4]
```diff
@@ -1,41 +1,158 @@
...
+    }: {
+      [LEGACY_PASS_HASH_STORE_KEY]?: unknown;
+      [SESSION_PAYLOAD_STORE_KEY]?: unknown;
+    } = await storage.get([LEGACY_PASS_HASH_STORE_KEY, SESSION_PAYLOAD_STORE_KEY]);
+
+    if (legacyPassHash !== undefined) {
+      await storage.remove(LEGACY_PASS_HASH_STORE_KEY);
+    }
+
+    if (!isVaultSessionPayloadV2(payload)) {
```
Comment (0xVoronov) [inline src/lib/temple/back/vault/session-store.ts]: Let's cover this branch with tests

File: src/lib/temple/back/vault/session-store.ts [hunk 5]
```diff
@@ -1,41 +1,158 @@
...
   }
 };
+
+function getSessionStorage() {
+  return browser.storage.session as SessionStorageArea | undefined;
+}
+
+async function restrictSessionAccess(storage: SessionStorageArea) {
+  try {
+    await storage.setAccessLevel?.({ accessLevel: 'TRUSTED_CONTEXTS' });
```
Comment (0xVoronov) [inline src/lib/temple/back/vault/session-store.ts:151]: Do we have a test for it?

File: src/lib/temple/back/main.ts [hunk 6]
```diff
@@ -1,21 +1,20 @@
 import browser, { Runtime } from 'webextension-polyfill';
```
Comment (0xVoronov) [inline src/lib/temple/back/main.ts:1]: ⚠️ CAN BE COVERED ON A SEPARATE PR LATER:

1. The SEC-06 attack integration test. Capture the browser.runtime.onConnect listener, connect a fake relay port (sender.tab, HTTPS URL), and verify that SignRequest / ConfirmationRequest{confirmed:true} result in an Unauthorized intercom request source error, while PageRequest is allowed through.
This will require exporting processRequest from main.ts (or moving it to a separate module).

2. Tests for Actions.sign / sendOperations: mock intercom.getPortInfo -> relay/unknown port -> throw before any vault.sign call; ExtensionUi port + ConfirmationRequest from a *different* port -> ignored; from the same port -> signed. This covers both reqPort === port and the repeated assertion.

3. Test for requestConfirm in dapp.ts: DAppGetPayloadRequest from an ExtensionUi port (not ConfirmUi) -> Unauthorized dApp confirmation source; from ConfirmUi -> payload; a confirmation from a port different from knownPort -> ignored.

4. Manifest snapshot test (covers regressions in both SEC-10 and SEC-03): content_scripts should contain exactly the content script (all_frames: false) + keepalive, and the WAR should not contain misc/ad-banners/*. Cheap and protects both “untestable” SECs.

5. Five negative branches in validateAuthChallengeForSigning (expired / TTL > 10 min / >1024 / bad nonce / non-ISO) - straightforward table-driven tests.

6. Session store: malformed payload -> removed + undefined; Vault.setup with an incorrect password does not write to the session; lock() calls forgetSession.

**Review by 0xVoronov — CHANGES_REQUESTED** [review]

File: src/lib/temple/back/vault/session-store.ts (hunk 3 above)
Comment (AlexBelz123) [reply src/lib/temple/back/vault/session-store.ts:55]: Reworked vault session recovery to v3: session payload stays in browser.storage.session, the non-extractable wrapping key is stored separately via IndexedDB in session-wrapping-key-store.ts, and recovery returns a non-extractable PBKDF2 CryptoKey

File: src/lib/temple/back/intercom-permissions.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+const RELAY_ALLOWED_REQUEST_TYPES = new Set<TempleMessageType>([
+  TempleMessageType.PageRequest,
+  TempleMessageType.Acknowledge
+]);
+
+export const isTempleRequestAllowedForPortInfo = (req: TempleRequest, portInfo: IntercomPortInfo) => {
+  switch (portInfo.kind) {
+    case IntercomPortKind.ContentScriptRelay:
+      return RELAY_ALLOWED_REQUEST_TYPES.has(req.type) && (portInfo.frameId == null || portInfo.frameId === 0);
```
Comment (TristanAllaire) [inline src/lib/temple/back/intercom-permissions.ts]: frameId can't be null so keeping only portInfo.frameId === 0 is enough.

File: src/lib/temple/back/intercom-permissions.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  }
+};
+
+export const assertTempleRequestAllowedForPortInfo = (req: TempleRequest, portInfo: IntercomPortInfo) => {
+  if (!isTempleRequestAllowedForPortInfo(req, portInfo)) {
+    throw new Error('Unauthorized intercom request source');
+  }
+};
+
+export const assertTrustedUiPortInfo = (portInfo?: IntercomPortInfo) => {
```
Comment (TristanAllaire) [inline src/lib/temple/back/intercom-permissions.ts]: assertTrustedUiPortInfo is exported but has no callers anywhere: dead trust-check. Remove it, or wire it in so we're not shipping an unused auth helper.

File: src/lib/temple/back/main.ts [hunk 9]
```diff
@@ -1,24 +1,24 @@
...
-import { AnalyticsEventCategory } from '../analytics-types';
-
 import * as Actions from './actions';
 import * as Analytics from './analytics';
 import { intercom } from './defaults';
+import { assertTempleRequestAllowedForPortInfo } from './intercom-permissions';
 import { store, toFront } from './store';
 
 const frontStore = store.map(toFront);
+const isE2eResetEnabled = process.env.NODE_ENV !== 'production';
```
Comment (TristanAllaire) [inline src/lib/temple/back/main.ts]: This gate relies on NODE_ENV !== 'production'. Please confirm the prod webpack build sets NODE_ENV=production: if it's ever unset, this full-storage-wipe (ResetRequest) handler ships live.

File: src/lib/temple/back/vault/session-wrapping-key-store.ts [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+  await Promise.all(expiredIds.map(deleteSessionWrappingKey));
+}
+
+function getIndexedDb() {
+  if (typeof indexedDB !== 'undefined') return indexedDB;
+  return undefined;
+}
+
+function isTestEnv() {
```
Comment (TristanAllaire) [inline src/lib/temple/back/vault/session-wrapping-key-store.ts:80]: The in-memory memoryRecords fallback only triggers on NODE_ENV === 'test'. Worth guaranteeing prod never reaches it, otherwise a build with NODE_ENV unset would keep wrapping keys in a JS Map instead of IndexedDB.

File: src/lib/temple/back/vault/session-wrapping-key-store.ts [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+  if (!indexedDb) {
+    if (isTestEnv()) memoryRecords.clear();
+    return;
+  }
+
+  await runRequest('readwrite', store => store.clear());
+}
+
+export async function clearExpiredSessionWrappingKeys(now = Date.now()) {
+  const indexedDb = getIndexedDb();
```
Comment (TristanAllaire) [inline src/lib/temple/back/vault/session-wrapping-key-store.ts:59]: clearExpiredSessionWrappingKeys() has no callers. Either call it on recovery/startup to GC expired keys from IndexedDB, or drop it. (Low impact since saveSessionPassHash already clears prior keys, but it's dead as written.)

File: src/lib/temple/back/vault/session-store.ts [hunk 12]
```diff
@@ -1,41 +1,195 @@
...
+  deleteSessionWrappingKey,
+  getSessionWrappingKey,
+  saveSessionWrappingKey
+} from './session-wrapping-key-store';
+
+const LEGACY_PASS_HASH_STORE_KEY = '@Vault:session.passHash';
+const SESSION_PAYLOAD_STORE_KEY_V2 = '@Vault:session.v2';
+const SESSION_PAYLOAD_STORE_KEY_V3 = '@Vault:session.v3';
+const SESSION_PAYLOAD_VERSION = 3;
+const SESSION_PAYLOAD_TTL_MS = 12 * 60 * 60 * 1000;
```
Comment (TristanAllaire) [inline src/lib/temple/back/vault/session-store.ts:17]: Session recovery window is 12h (SESSION_PAYLOAD_TTL_MS). Please confirm this matches the product auto-lock policy. The worker can silently recover the unlocked vault for this whole window.

File: src/mavryk/api/client.ts
Comment (TristanAllaire) [inline src/mavryk/api/client.ts]: isAuthRefreshCandidate matches endpoints with substring .includes(). Any future protected route whose path contains /auth/refresh (etc.) would silently lose its Authorization header. Consider matching the endpoint/pathname exactly.

File: src/mavryk/api/auth.ts [hunk 13]
```diff
@@ -110,27 +109,35 @@ export async function refreshAuthTokens(params: AuthRefreshRequest = {}) {
 }
 
 export async function logoutAuth(params: AuthRefreshRequest = {}) {
```
Comment (TristanAllaire) [inline src/mavryk/api/auth.ts:111]: Heads-up: with no params, logoutAuth now clears and revokes refresh tokens for all wallets/networks, not just the current one. If that's the intended behaviour on lock, good . Just flagging so it isn't "fixed" back to single-scope later.

Comment (AlexBelz123) [reply src/mavryk/api/auth.ts:111]: That one was already correct, so I did not change the logic.
logoutAuth() with no params intentionally clears/revokes all stored refresh tokens across wallets/networks. I added a code comment so it does not get “fixed” back to single-scope later

File: src/lib/temple/back/vault/session-wrapping-key-store.ts [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+export type VaultSessionWrappingKeyRecord = {
+  id: string;
+  createdAt: number;
+  expiresAt: number;
+  key: CryptoKey;
+};
+
+const memoryRecords = new Map<string, VaultSessionWrappingKeyRecord>();
+
+export async function saveSessionWrappingKey(record: VaultSessionWrappingKeyRecord) {
```
Comment (0xVoronov) [inline src/lib/temple/back/vault/session-wrapping-key-store.ts:14]: ⚠️ CAN BE COVERED IN A SEPARATE PR:

Let's cover the real IndexedDB path of the wrapping-key store with tests and then remove the test fallback from production code.

jsdom (jest) has no global `indexedDB`, so the session-store tests only ever exercise the `!indexedDb` → `isTestEnv()` → `memoryRecords` branch. The real production path - `openDb`/`runRequest` (store creation in `onupgradeneeded`, transaction lifecycle, error/abort handling, `db.close`) - is never executed in CI. In prod it's the exact opposite: `indexedDB` always exists in extension contexts, so the Map branch is unreachable - it's pure test infrastructure compiled into the production bundle. The only case where prod could hit this fork (no IDB + `NODE_ENV !== 'test'`) throws, which is exactly the fail-closed behavior we want to keep (already covered by your new test 👍).

Suggested approach:

1. Wire `fake-indexeddb` into jest (`setupFiles: ['fake-indexeddb/auto']`) and cover the store mechanics: `onupgradeneeded` creates the object store; put/get/delete/clear round-trip; `clearExpiredSessionWrappingKeys` removes only expired records; transaction error/abort → rejection.
2. Heads-up: `fake-indexeddb@6` does NOT round-trip `CryptoKey` - `put` succeeds, but `get` silently returns a plain object without the prototype, so the key fails on first use (`TypeError: Key is not of type 'CryptoKey'`). For the `key` field either patch the structured clone in jest.setup to pass `CryptoKey` through by reference, or verify the real round-trip at the e2e level (unlock → SW restart → wallet recovers without a password; after `storage.session` is cleared → password required).
3. Once tests run against fake IDB, remove `memoryRecords`/`isTestEnv()` from the module and keep only the throw.

**Review by 0xVoronov — APPROVED** [review]
